add LA::d::BlockVector:: multivector_inner_product()

davydden opened the pull request:
reasons: (i) convenience (ii) a single global reduction as opposed to N*M reductions of each (i,j) products.

davydden commented:
/run-tests

kronbichler commented:
How about choosing putting name like `multivector_outer_product`? This is a product that treats the `BlockVector` as a `MultiVector` with individual vectors representing different things, whereas the standard use of the class and all other name conventions are for different blocks of a single solution field where the inner product and things like that go over all block components. Thus, I think that a name that clearly indicates that this assumes a different interpretation as the standard one is better.

Second reason: I think that `scalar_product` is misleading in this case because one expects to have a scalar as a result (at least in the German-dominated part of English math literature) or a scalar as the factor by which one multiplies with.

davydden commented:
> multivector_outer_product

it's certainly not outer product, otherwise you would be getting some matrices based on vectors, but if you prefer I can rename it to `multivector_inner_product`. The point is, it's inner products between each pair of blocks for two vectors. So from that perspective `scalar_product` is also ok, it is just not the product between the whole thing, but scalar product between each pair of blocks.

kronbichler commented on include/deal.II/lac/la_parallel_block_vector.h:
```diff
@@ -581,6 +596,61 @@ namespace LinearAlgebra
+    }
+
+
+#endif
```
I thinks this function is big enough to go into the `.templates.h` file with instantiations for the appropriate other scalar type and full matrix type. But of course the question is how many different matrix types and results types we want to support.

davydden replied:
at least `FullMatrix` and `LAPACKFullMatrix` with doubles. I can do that, anything else can be used by including `.templates.h`.

davydden commented:
@kronbichler let me see how Trilinos name it for multivectors....

davydden commented:
nevermind, can't find it...

how about `multivector_inner_product` ?

kronbichler commented:
My reasoning was that it is an *outer* product over the columns of the multivector, but you are right that it's not descriptive enough because we have an *inner* product around the rows in the multivector. So yes, `multivector_inner_product` sounds good to me.

davydden commented:
I addressed your comments. I will split the changes into two commits shortly...

p.s. the closest I found in Trilinos is named `MvTransMv` in various places: https://trilinos.org/docs/dev/packages/anasazi/doc/html/classAnasazi_1_1MultiVecTraits.html#a9f20552d4ee9c55d559f20576bb24ce0 and https://trilinos.org/docs/dev/packages/belos/doc/html/classBelos_1_1MultiVec.html#af62a827951530e121969e1974efeb90c

davydden commented:
@drwells may not like another MPI::sum() here 😄

bangerth commented on include/deal.II/lac/la_parallel_block_vector.h:
```diff
@@ -448,6 +448,21 @@ namespace LinearAlgebra
+       * Obviously, this function can only be used if each block in this
+       * object and @p V are of the same size.
+       *
+       * @note internally, a single global reduction will be called to
```
-> Internally

bangerth commented on include/deal.II/lac/la_parallel_block_vector.templates.h:
```diff
@@ -779,6 +779,56 @@ namespace LinearAlgebra
+                                                   const bool symmetric) const
+    {
+      const unsigned int m = this->n_blocks();
+      const unsigned int n = V.n_blocks();
```
Shouldn't you assert that `n==m`?

davydden replied:
nope, only in case it's symmetric which happens below. Otherwise it's perfectly fine to have non-square matrices.

bangerth commented on include/deal.II/lac/la_parallel_block_vector.h:
```diff
@@ -448,6 +448,21 @@ namespace LinearAlgebra
+       * Calculate scalar product between each block of this vector and @p V
+       * and store the result in a full matrix @p matrix. If @p symmetric is
+       * <code>true</code>, it is assumed that inner product results in a
+       * square symmetric matrix and almost half of the scalar products can be avoided.
```
I needed to read the implementation to find out what you are saying here with the matrix. I think it would be useful to make it explicitly that
```
  This function computes the result by forming $A_{ij}=U_i \cdot V_j$ where $U_i$ and $V_i$ indicate the
  $i$th block (not element!) of $U$ and the $j$th block of $V$, respectively.
```

davydden commented:
> Are you so concerned about performance that you don't want to just run a double loop and compute the regular dot product between vectors? The only difference is that you'll have N^2 communication 
 steps instead of 1 communication step. Everything is the same.

yes, I am interested in cases where you have hundreds of blocks on each side and the number of products is on the order of tens/hundreds of thousands.

bangerth commented:
Ah, ok.

What are you thinking about returning the computed matrix by-value, rather than via an argument? Surely that's cheaper than the communication, and I think that's the direction we're generally going :-)

davydden commented:
> What are you thinking about returning the computed matrix by-value, rather than via an argument? Surely that's cheaper than the communication, and I think that's the direction we're generally going :-)

I am ok with that but one won't be able to guess the template parameter from arguments, but that's fine with me.

bangerth commented:
I think it should just be a `FullMatrix` of the type that the scalar product of the individual blocks returns. So something like `FullMatrix<decltype(block(0) * block(0))>` if you wanted to be completely generic, but it should be obvious what to replace the `decltype` by.

davydden commented:
> I think it should just be a FullMatrix 

that's a limitation. It should certainly support `LAPACKFullMatrix`, as this is also a highly used alternative with its own advantages.

davydden commented:
> scalar product of the individual blocks returns

In practice, you would not use a different number fro two multivectors, so having a single `NumberType` should be perfectly enough. At least, that's my intention in this PR. If someone wants to extend it to mixed types later, that's fine.

bangerth commented:
Yes, I think `NumberType` is exactly what you need. That's of course also exactly what the scalar product should return.

Re `LAPACKFullMatrix` -- can't you just convert from one matrix type to the other at the caller's site? If `LAPACKFullMatrix` does not have an `operator=`, then maybe we should add one. We already have so many template instantiations...

davydden commented:
> can't you just convert from one matrix type to the other at the caller's site?

not directly, but then it defeats the whole purpose of returning by value with move constructor, right? Because you would essentially kill whatever is created inside and copy the result into an object of different class outside of this function.

> We already have so many template instantiations...

I would argue this PR is not the place to hunt this problem :smile:

davydden commented:
a bit more to the discussion: FullMatrix and LAPACKFullMatrix are not very interchangeable as they currently support very different number types. Not that BlockVector's are tested with complex numbers, but still...

bangerth commented:
That's fair.

I'm not going to stand in the way if you feel that that's an important issue to you. But I do think that copying a matrix is likely going to cost far less than (i) the operations and communication you just did, and (ii) the cost of whatever you plan on doing with the matrix.

I could imagine that there are ways to implement `LAPACKFullMatrix::operator= (FullMatrix &&)` (i.e., a move operator; alternatively, a move constructor) efficiently. But that's besides the point.

davydden commented:
> But I do think that copying a matrix is likely going to cost far less than (i) the operations and communication you just did, and (ii) the cost of whatever you plan on doing with the matrix.

that certainly true and we do have 

```
LAPACKFullMatrix< number > & LAPACKFullMatrix< number >::operator= (const FullMatrix<number2> & M) 
```
but there is something which is slightly inconvenient here: 

When we copy a full matrix into LAPACK we don't know whether the input corresponds to a symmetric matrix or not (memory allocation is the same). After copying internally LAPACK matrix will store a property that it contains a general function. So the users would have to set it to symmetric manually after calling the function. This can be avoided / automated if we either (a) return LAPACKFullMatrix or (b) take it by reference. In both cases, internally one could set the property accordingly. 

This is not a deal-breaker for sure, but still quite significant issue which is easy to overlook... 

p.s. FullMatrix does not know anything about symmetry, so we don't need to set extra states or something.

> I could imagine that there are ways to implement LAPACKFullMatrix::operator= (FullMatrix &&) (i.e., a move operator; alternatively, a move constructor) efficiently. But that's besides the point.

but they store data in different format (row and column major), I am not sure one can do anything about it...

bangerth commented:
All good points. Let's not worry about it then!

davydden commented:
i will push a quick patch soon to make sure LAPACKFullMatrix has Property set correctly inside the function...

davydden commented:
I update the PR to set LAPACKSupport::Property accordingly (it's not super important, but it's the right thing to do IMO).

davydden commented:
@kronbichler @bangerth ping.

bangerth commented on include/deal.II/lac/la_parallel_block_vector.h:
```diff
@@ -448,6 +448,24 @@ namespace LinearAlgebra
+       * Calculate the scalar product between each block of this vector and @p V
+       * and store the result in a full matrix @p matrix. This function
+       * computes the result by forming $A_{ij}=U_i \cdot V_j$ where $U_i$
+       * and $V_i$ indicate the $i$th block (not element!) of $U$ and the
```
-> V_j

bangerth commented on include/deal.II/lac/la_parallel_block_vector.h:
```diff
@@ -448,6 +448,24 @@ namespace LinearAlgebra
+       * square symmetric matrix and almost half of the scalar products can be avoided.
+       *
+       * Obviously, this function can only be used if each block in this
+       * object and @p V are of the same size.
```
-> all blocks of both vectors are ...

bangerth commented on include/deal.II/lac/la_parallel_block_vector.h:
```diff
@@ -448,6 +448,24 @@ namespace LinearAlgebra
+       * accumulate scalar product between locally owned degrees of freedom.
+       */
+      template <typename FullMatrixType>
+      void multivector_inner_product(FullMatrixType &matrix, const BlockVector<Number> &V, const bool symmetric = false) const;
```
can you break the arguments onto separate lines?
